### PR TITLE
Implement Meta Delete functionality

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This PR introduces `meta_delete` functionality and does some further refactoring to simplify code after patterns and common code between different methods became more clear.  

### Details - How are you making this change?  What are the effects of this change?

The [Meta Delete functionality](https://github.com/memcached/memcached/blob/8203c934bac24ff9f54e837d618c2521ef826a64/doc/protocol.txt#L797), introduced here covers associated CAS semantics, key invalidation using the `I` flag, and tombstoning capability when using the `x` flag.

Benchmarks haven't been added because they would be skewed by the need to re-add keys to avoid spamming on a non-existent key.

### CODE REVIEW

File: src/parser/meta_parser.rs [hunk 1]
```diff
@@ -178,80 +188,27 @@ fn parse_meta_get_data_value(buf: &[u8]) -> IResult<&[u8], MetaResponse> {
...
+        // match arm for "MN\r\n" response
+        MetaResponse::Status(Status::NoOp) => Ok((input, MetaResponse::Status(Status::NoOp))),
+        // match arm for "HD", "NS", "EX" & "NF" responses
+        MetaResponse::Status(s) => process_meta_response_without_data_payload(input, s),
+        _ => Err(nom::Err::Error(nom::error::Error::new(
+            input,
+            nom::error::ErrorKind::Eof,
+        ))),
+    }
+}
```
Comment (mrattle) [inline src/parser/meta_parser.rs:202]: These can all be combined into one arm because they are semantically the same, only the response code being returned is different.

File: src/proto/meta_protocol.rs [hunk 2]
```diff
@@ -84,28 +98,16 @@ impl MetaProtocol for Client {
...
-        } else {
-            match self.drive_receive(parse_meta_get_response).await? {
-                MetaResponse::Status(Status::NotFound) => Ok(None),
-                MetaResponse::Status(s) => Err(s.into()),
-                MetaResponse::Data(d) => d
-                    .map(|mut items| {
-                        let item = items.remove(0);
-                        Ok(item)
-                    })
-                    .transpose(),
```
Comment (mrattle) [inline src/proto/meta_protocol.rs:107]: This conditional wasn't actually required in this case, the match arms can be condensed together for the same function.

File: src/parser/meta_parser.rs [hunk 3]
```diff
@@ -400,6 +357,18 @@ fn construct_meta_value_from_flag_array(
...
+    // let data = if let Some(d) = data {
+    //     if d.is_empty() {
+    //         None
+    //     } else {
+    //         Some(d.to_vec())
+    //     }
+    // } else {
+    //     None
+    // };
+
```
Comment (mrattle) [inline src/parser/meta_parser.rs:371]: Having the MetaValue's data attribute mapped to vec![] after tombstoning (clearing the key's associated value with the x flag) doesn't seem right, we might need a small parser change to have this mapped to None instead.

Basically what's happening is that the parser currently expects a data block as part of a VA response from a meta get, but after removing the value there is simply nothing there.

File: src/parser/meta_parser.rs [hunk 4]
```diff
@@ -178,80 +188,49 @@ fn parse_meta_get_data_value(buf: &[u8]) -> IResult<&[u8], MetaResponse> {
...
+    }
+}
 
-            let meta_value = construct_meta_value_from_flag_array(
-                meta_values_array,
-                None,
-                Some(Status::NotFound),
-            )
-            .map_err(|_| nom::Err::Failure(nom::error::Error::new(buf, Fail)))?;
+// example meta_delete command sent to memcached server:
```
Comment (nickamorim) [inline src/parser/meta_parser.rs:204]: Is this just leftover GPT shtuff?

File: src/parser/meta_parser.rs (hunk 3 above)
Comment (nickamorim) [reply src/parser/meta_parser.rs:393]: > change to have this mapped to None instead

This sounds like it makes sense

**Review by nickamorim — APPROVED** [review]
